yarn-plugin/linkers/pnp.ts

```diff
@@ -0,0 +1,65 @@
+/* istanbul ignore file */
+// this file is covered by CLI tests
+
+import { getPnpPath } from "@yarnpkg/plugin-pnp";
+import type { Package, Project } from "@yarnpkg/core";
+import { structUtils } from "@yarnpkg/core";
+import type { PortablePath } from "@yarnpkg/fslib";
+import { VirtualFS } from "@yarnpkg/fslib";
+// import { getLibzipSync, ZipOpenFS } from "@yarnpkg/libzip";
+import { getArchitectureSet } from "./utils";
+
+/**
+ * Get package path with `pnp` linker for a given Yarn project and package
+ *
+ * @param {Project} project - Yarn project
+ * @param {Package} pkg - Yarn package
+ * @returns {Promise<PortablePath | null>} Package path
+ */
+export const getPackagePath = async (
+  project: Project,
+  pkg: Package,
+): Promise<PortablePath | null> => {
+  makePnPApi(project);
+
+  if (!structUtils.isPackageCompatible(pkg, getArchitectureSet())) return null;
+
+  const locator = structUtils.convertPackageToLocator(pkg);
+  const pnpLocator = {
+    name: structUtils.stringifyIdent(locator),
+    reference: locator.reference,
+  };
+
+  const packageInformation = pnpApi.getPackageInformation(pnpLocator);
+  if (!packageInformation) return null;
+
+  const { packageLocation } = packageInformation;
+  return packageLocation;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+let pnpApi: any;
+
+/**
+ * Cache PnP API from `.pnp.cjs` (or similar), if it has not already been cached
+ *
+ * @param {Project} project - Yarn project
+ * @returns {void}
+ */
+const makePnPApi = (project: Project): void => {
+  if (!pnpApi) {
+    // use `module.require` so webpack leaves this alone
+    pnpApi = module.require(getPnpPath(project).cjs);
+  }
+};
+
+/**
+ * Instantiate the virtual file system for reading package files in PnP
+ */
+export const fs = new VirtualFS({
+  // TODO: Fix this. When using the plugin, it throws the type error: pe.ZipOpenFS is not a constructor
```

Comment:
Maybe make the PR "draft" because there are still a few TODOs?

Reply:
The TODOs are meant to be implemented in later PRs. The idea of this PR is only to create a rough and incomplete, but working proof of concept for the yarn plugin.
We don't have draft PRs activated yet for this repository.